Updates `dirs` builtin

timetoplatypus opened the pull request:
- Switches the `dirs` implementation over to the tilde-prefix format when printing the directory stack.
- Fixes a few bugs including when the user invokes Pushd/Popd/Dirs with path aliases such as `.` and `..`
- Adds more spec tests for `dirs`

andychu commented on core/builtin.py:
```diff
@@ -604,28 +604,51 @@ def Cd(argv, mem, dir_stack):
+LONG_FORM = 2
 
-def _PrintDirStack(dir_stack, style):
+def _PrintDirStack(dir_stack, mem, style, form):
```
I think the interface to this function can be simplified.  Why not just pass an optional `home_dir=None` arg?  If it's not None, then do the translation.

We don't need `mem` or `form`.  That can be pushed to the outer levels.

andychu commented on core/builtin.py:
```diff
@@ -604,28 +604,51 @@ def Cd(argv, mem, dir_stack):
+  if style == WITH_LINE_NUMBERS:
     for i, entry in enumerate(dir_stack.Iter()):
-      print('%2d  %s' % (i, entry))
+      if form == TILDE_PREFIX_FORM:
```
Likewise you can get rid of the `if` here with a function like

`_FormatDir(entry, home_dir=None)`

andychu commented on core/builtin.py:
```diff
@@ -604,28 +604,51 @@ def Cd(argv, mem, dir_stack):
+      for i, entry in enumerate(dir_stack.Iter()):
+        # Don't add a space after printing the path if it's
+        # the last entry being printed.
+        sys.stdout.write(_TranslateHomePrefix(entry, mem) + \
```
This looks equivalent to `' '.join()`

andychu commented on core/builtin.py:
```diff
@@ -604,28 +604,51 @@ def Cd(argv, mem, dir_stack):
+def _TranslateHomePrefix(dir_name, mem):
+  val = mem.GetVar('HOME')
+  if val.tag == value_e.Str:
+    return dir_name.replace(val.s, '~', 1)
```
I think you have to check if the value starts with the home_dir.  `replace()` will replace it in the middle, which isn't correct.

timetoplatypus replied:
Good point, I can add a check for that.

andychu commented on core/state.py:
```diff
@@ -263,6 +263,8 @@ def Iter(self):
     return reversed(self.stack)
 
+  def Size(self):
+    return len(self.stack)
```
not needed if you follow my suggestions

andychu commented on core/cmd_exec.py:
```diff
@@ -275,13 +275,13 @@ def _RunBuiltin(self, builtin_id, argv):
 
     elif builtin_id == EBuiltin.PUSHD:
-      status = builtin.Pushd(argv, self.dir_stack)
+      status = builtin.Pushd(argv, self.mem, self.dir_stack)
```
Although it's a little redundant, I think all three functions should just take `self.mem.GetVar('HOME')`

In general it's better to have a narrower interface, and not depend on all of `Mem()` when you only need to read a single variable.

timetoplatypus replied:
True, it does only need one variable. I can change it to just take the value of $HOME specifically

andychu commented:
Not sure if you are done with this, but it looks like the `replace()` call still needs to be fixed.

I get a notification from Github when you push.  In the future I will not look unless you explicitly respond with some comments.  Just so the workflow is clear.

timetoplatypus commented:
Yep, I'm still working on it.

Sounds good, I'll post on whatever thread we're working on when the code is ready for review.

timetoplatypus commented:
Do you think that situation is worth writing a test for? (Where a path may have "/home/<username>" in the middle of it, so that shouldn't be translated to a tilde)

timetoplatypus commented:
I fixed the `replace()` issue by checking if the input starts with the home directory path. Let me know if you'd like a spec test for that too. Otherwise, the code should be ready for another review.

andychu commented on core/builtin.py:
```diff
@@ -604,28 +604,35 @@ def Cd(argv, mem, dir_stack):
 
-def Pushd(argv, dir_stack):
+def _FormatDir(dir_name, home_dir):
+  if home_dir and home_dir.tag == value_e.Str and dir_name.startswith(home_dir.s):
```
I think there is still a bug here, where

```
HOME=/home/andy
```

and

one of the entries is `/home/andyc`.  In that case it doesn't make sense to print `~c`.

So you have to handle at least two cases, and yes there should be tests for these.

timetoplatypus replied:
Ahh, yes I missed that case. I'll write up some tests and figure out how to handle that

andychu commented on core/builtin.py:
```diff
@@ -604,28 +604,35 @@ def Cd(argv, mem, dir_stack):
 
   elif style == SINGLE_LINE:
-    print(' '.join(dir_stack.Iter()))
+    print(' '.join(_FormatDir(entry, home_dir) \
```
Don't need `\`.  In Python wrapping is assumed between `()`.  Also indent continuation lines by 4.

timetoplatypus commented:
I added spec tests for those aforementioned cases.

For the bug you mentioned where $HOME appears as a substring of a different legitimate path, I changed the code so that when it does the comparison using `.startswith()` the path has to also end with another forward slash in order to match.

Can you take a look and let me know if I missed anything?

andychu commented:
What about the case where `$PWD` is exactly `$HOME` ?

timetoplatypus commented:
Good point. I've added a spec test for that case and fixed it in code by checking if the directory name passed in is exactly the home directory passed in. If it is, it'll short circuit and proceed with the tilde replacement. Otherwise, it'll continue to check for whether the directory name passed in starts with the home directory.

andychu commented:
OK thanks.  I split the dirs tests out into their own file because only bash and zsh implement it.

Generally everything should be green and yellow, and dash and mksh were all red.

There was one case that zsh was red on too.  If there is red, then I mark it "OK" and with the differing stdout.

https://github.com/oilshell/oil/commit/4a12881dfb6d56295001201136c33019217a0c

If anything about the spec tests don't make sense let me know.
